tests/integration/backward_compatible/sql_test.py

```diff
@@ -0,0 +1,590 @@
+                pass
+
+    def test_execute_with_mismatched_params_when_params_has_more(self):
+        self._populate_map()
+        result = self.client.sql.execute("SELECT * FROM %s WHERE this > ?" % self.map_name, 5, 6)
+
+        with self.assertRaises(HazelcastSqlError):
+            for _ in result:
+                pass
+
+    def test_execute_statement(self):
+        entry_count = 12
+        self._populate_map(entry_count, str)
+        statement = SqlStatement("SELECT this FROM %s" % self.map_name)
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(
+            self,
+            [str(i) for i in range(entry_count)],
+            [row.get_object_with_index(0) for row in result],
+        )
+
+    def test_execute_statement_with_params(self):
+        entry_count = 20
+        self._populate_map(entry_count, lambda v: Student(v, v))
+        statement = SqlStatement(
+            "SELECT age FROM %s WHERE height = CAST(? AS REAL)" % self.map_name
+        )
+        statement.add_parameter(13.0)
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(self, [13], [row.get_object("age") for row in result])
+
+    def test_execute_statement_with_mismatched_params_when_sql_has_more(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT * FROM %s WHERE __key > ? AND this > ?" % self.map_name)
+        statement.parameters = [5]
+        result = self.client.sql.execute_statement(statement)
+
+        with self.assertRaises(HazelcastSqlError):
+            for _ in result:
+                pass
+
+    def test_execute_statement_with_mismatched_params_when_params_has_more(self):
+        self._populate_map()
+        statement = SqlStatement("SELECT * FROM %s WHERE this > ?" % self.map_name)
+        statement.parameters = [5, 6]
+        result = self.client.sql.execute_statement(statement)
+
+        with self.assertRaises(HazelcastSqlError):
+            for _ in result:
+                pass
+
+    def test_execute_statement_with_timeout(self):
+        entry_count = 100
+        self._populate_map(entry_count, lambda v: Student(v, v))
+        statement = SqlStatement("SELECT age FROM %s WHERE height < 10" % self.map_name)
+        statement.timeout = 100
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(
+            self, [i for i in range(10)], [row.get_object("age") for row in result]
+        )
+
+    def test_execute_statement_with_cursor_buffer_size(self):
+        entry_count = 50
+        self._populate_map(entry_count, lambda v: Student(v, v))
+        statement = SqlStatement("SELECT age FROM %s" % self.map_name)
+        statement.cursor_buffer_size = 3
+        result = self.client.sql.execute_statement(statement)
+
+        six.assertCountEqual(
+            self, [i for i in range(entry_count)], [row.get_object("age") for row in result]
+        )
```

Comment:
You can add a spy to fetch calls and assert its count here

Reply:
This is an integration test where we use the real code, not the mocked one. I am not planning to add mocking to this test but as long as the server is not faulty (respecting the cursor buffer size), fetch will be called.

Reply:
You don't need to mock to spy in Node.js. Does python have such feature? See spies in sinon package.